File: smoke_tests/scala/jndi_injection.scala
// JNDI Injection vulnerabilities in Scala
package com.example.security

import javax.naming.{Context, InitialContext}
import java.util.Hashtable

class JndiInjectionVulnerabilities {

  // Test 1: Direct JNDI lookup with user input
  def lookupResource(resourceName: String): Any = {
    // VULNERABLE: User-controlled JNDI name
    val ctx = new InitialContext()
    ctx.lookup(resourceName)
  }

  // Test 2: JNDI lookup with URL
  def lookupByUrl(jndiUrl: String): Any = {
    // VULNERABLE: URL from user (can be rmi://, ldap://)
    val env = new Hashtable[String, String]()
    env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.rmi.registry.RegistryContextFactory")
    env.put(Context.PROVIDER_URL, jndiUrl)
    val ctx = new InitialContext(env)
    ctx.lookup("name")
  }

  // Test 3: DataSource lookup
  def getDataSource(dsName: String): javax.sql.DataSource = {
    // VULNERABLE: DataSource name from user
    val ctx = new InitialContext()
    ctx.lookup(s"java:comp/env/jdbc/$dsName").asInstanceOf[javax.sql.DataSource]
  }

  // Test 4: JMS destination lookup
  def getDestination(destName: String): Any = {
    // VULNERABLE: Destination from user
    val ctx = new InitialContext()
    ctx.lookup(s"jms/$destName")
  }

  // Test 5: Environment lookup
  def getEnvValue(envName: String): Any = {
    // VULNERABLE: Environment name from user
    val ctx = new InitialContext()
    ctx.lookup(s"java:comp/env/$envName")
  }

  // Test 6: Remote object lookup
  def getRemoteObject(serviceName: String): Any = {
    // VULNERABLE: Service name injection
    val env = new Hashtable[String, String]()
    env.put(Context.PROVIDER_URL, "rmi://localhost:1099")
    val ctx = new InitialContext(env)
    ctx.lookup(serviceName)
  }

  // Test 7: LDAP JNDI lookup
  def ldapLookup(ldapUrl: String, baseDn: String): Any = {
    // VULNERABLE: LDAP URL from user
    val env = new Hashtable[String, String]()
    env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory")
    env.put(Context.PROVIDER_URL, ldapUrl)
    val ctx = new InitialContext(env)
    ctx.lookup(baseDn)
  }

  // Test 8: Log4j style JNDI (Log4Shell pattern)
  def logMessage(message: String): Unit = {
    // VULNERABLE: Message could contain ${jndi:ldap://...}
    val logger = org.slf4j.LoggerFactory.getLogger(getClass)
    logger.info(message)
  }

  // Test 9: Object factory lookup
  def getObjectFromFactory(factoryName: String): Any = {
    // VULNERABLE: Factory name from user
    val ctx = new InitialContext()
    ctx.lookup(s"java:global/$factoryName")
  }

  // Test 10: Composite name lookup
  def compositeLookup(parts: List[String]): Any = {
    // VULNERABLE: Parts from user
    val ctx = new InitialContext()
    val name = new javax.naming.CompositeName()
    parts.foreach(name.add)
    ctx.lookup(name)
  }

  // Test 11: EJB lookup
  def lookupEjb(ejbName: String): Any = {
    // VULNERABLE: EJB name from user
    val ctx = new InitialContext()
    ctx.lookup(s"java:global/app/$ejbName")
  }

  // Test 12: Play Framework configuration lookup
  def playConfigLookup(key: String): Any = {
    // VULNERABLE: Config key from user
    val ctx = new InitialContext()
    ctx.lookup(s"java:comp/env/config/$key")
  }
}

File: smoke_tests/kotlin/jndi_injection.kt
// JNDI Injection vulnerabilities in Kotlin
package com.example.security

import javax.naming.Context
import javax.naming.InitialContext
import java.util.Hashtable

class JndiInjectionVulnerabilities {

    // Test 1: Direct JNDI lookup with user input
    fun lookupResource(resourceName: String): Any? {
        // VULNERABLE: User-controlled JNDI name
        val ctx = InitialContext()
        return ctx.lookup(resourceName)
    }

    // Test 2: JNDI lookup with URL
    fun lookupByUrl(jndiUrl: String): Any? {
        // VULNERABLE: URL from user (can be rmi://, ldap://)
        val env = Hashtable<String, String>()
        env[Context.INITIAL_CONTEXT_FACTORY] = "com.sun.jndi.rmi.registry.RegistryContextFactory"
        env[Context.PROVIDER_URL] = jndiUrl
        val ctx = InitialContext(env)
        return ctx.lookup("name")
    }

    // Test 3: DataSource lookup
    fun getDataSource(dsName: String): javax.sql.DataSource? {
        // VULNERABLE: DataSource name from user
        val ctx = InitialContext()
        return ctx.lookup("java:comp/env/jdbc/$dsName") as? javax.sql.DataSource
    }

    // Test 4: JMS destination lookup
    fun getDestination(destName: String): Any? {
        // VULNERABLE: Destination from user
        val ctx = InitialContext()
        return ctx.lookup("jms/$destName")
    }

    // Test 5: Environment lookup
    fun getEnvValue(envName: String): Any? {
        // VULNERABLE: Environment name from user
        val ctx = InitialContext()
        return ctx.lookup("java:comp/env/$envName")
    }

    // Test 6: Remote object lookup
    fun getRemoteObject(serviceName: String): Any? {
        // VULNERABLE: Service name injection
        val env = Hashtable<String, String>()
        env[Context.PROVIDER_URL] = "rmi://localhost:1099"
        val ctx = InitialContext(env)
        return ctx.lookup(serviceName)
    }

    // Test 7: LDAP JNDI lookup
    fun ldapLookup(ldapUrl: String, baseDn: String): Any? {
        // VULNERABLE: LDAP URL from user
        val env = Hashtable<String, String>()
        env[Context.INITIAL_CONTEXT_FACTORY] = "com.sun.jndi.ldap.LdapCtxFactory"
        env[Context.PROVIDER_URL] = ldapUrl
        val ctx = InitialContext(env)
        return ctx.lookup(baseDn)
    }

    // Test 8: Log4j style JNDI (Log4Shell pattern)
    fun logMessage(message: String) {
        // VULNERABLE: Message could contain ${jndi:ldap://...}
        val logger = org.slf4j.LoggerFactory.getLogger(this::class.java)
        logger.info(message)
    }

    // Test 9: Object factory lookup
    fun getObjectFromFactory(factoryName: String): Any? {
        // VULNERABLE: Factory name from user
        val ctx = InitialContext()
        return ctx.lookup("java:global/$factoryName")
    }

    // Test 10: Composite name lookup
    fun compositeLookup(parts: List<String>): Any? {
        // VULNERABLE: Parts from user
        val ctx = InitialContext()
        val name = javax.naming.CompositeName()
        parts.forEach { name.add(it) }
        return ctx.lookup(name)
    }

    // Test 11: EJB lookup
    fun lookupEjb(ejbName: String): Any? {
        // VULNERABLE: EJB name from user
        val ctx = InitialContext()
        return ctx.lookup("java:global/app/$ejbName")
    }

    // Test 12: Connection factory lookup
    fun getConnectionFactory(cfName: String): Any? {
        // VULNERABLE: Factory name from user
        val ctx = InitialContext()
        return ctx.lookup("java:/ConnectionFactory/$cfName")
    }
}
